src/main/java/com/epam/aidial/core/service/ResourceService.java

```diff
@@ -246,44 +304,146 @@ public String getResource(ResourceDescription descriptor, boolean lock) {
 
-            if (result.exists && !overwrite) {
-                return null;
+            if (metadata != null) {
+                if (!overwrite) {
+                    return null;
+                }
+
+                etag.validate(metadata.getEtag());
             }
 
-            long updatedAt = time();
-            long createdAt = result.exists ? result.createdAt : updatedAt;
-            redisPut(redisKey, new Result(body, createdAt, updatedAt, false, true));
+            long updatedAt = System.currentTimeMillis();
+            Long createdAt = metadata == null ? (Long) updatedAt : metadata.getCreatedAt();
+            String newEtag = EtagBuilder.generateEtag(body);
+            Result result = new Result(body, newEtag, createdAt, updatedAt, contentType, (long) body.length, false);
+            redisPut(redisKey, result);
 
             ResourceEvent event = new ResourceEvent()
                     .setUrl(descriptor.getUrl())
                     .setAction(ResourceEvent.Action.UPDATE)
                     .setTimestamp(updatedAt);
 
-            if (!result.exists) {
+            if (metadata == null) {
                 event.setAction(ResourceEvent.Action.CREATE);
-                blobPut(blobKey, "", createdAt, updatedAt); // create an empty object for listing
+                String blobKey = blobKey(descriptor);
+                blobPut(blobKey, result.toStub()); // create an empty object for listing
             }
 
             topic.publish(event);
-            return new ResourceItemMetadata(descriptor).setCreatedAt(createdAt).setUpdatedAt(updatedAt);
+            return descriptor.getType() == ResourceType.FILE
+                    ? toFileMetadata(descriptor, result)
+                    : toResourceItemMetadata(descriptor, result);
         }
     }
 
+    public Future<FileMetadata> putFile(
+            ResourceDescription descriptor, Pipe<Buffer> pipe, EtagHeader etag, String contentType) {
+        if (descriptor.getType() != ResourceType.FILE) {
+            throw new IllegalArgumentException("Expected a file, got %s".formatted(descriptor.getType()));
+        }
+
+        Promise<Void> promise = Promise.promise();
+        BlobWriteStream writeStream = new BlobWriteStream(vertx, blobStore, descriptor, contentType);
+        try {
+            etag.validate(() -> getEtag(descriptor));
+            pipe.to(writeStream, promise);
+        } catch (HttpException e) {
+            promise.fail(e);
+        }
+
+        return promise.future().map(v -> {
+            MultipartUpload mpu = writeStream.getMpu();
+            if (mpu == null) {
+                log.info("Resource is too small for multipart upload, sending as a regular blob");
+                byte[] body = writeStream.getBytes();
+                if (body.length <= maxSize) {
+                    return (FileMetadata) putResource(descriptor, body, etag, contentType, true, true);
+                }
+
+                String blobKey = blobKey(descriptor);
```

Comment:
Should have same flow with flushing and etc

Reply:
Fixed.